hubblestack/extmods/audit/bexpr.py

```diff
@@ -0,0 +1,197 @@
+        - expr: NOT (docker_installed AND dockerd_running) OR find_docker_file
+
+Here we are evaluating the boolean expression given in the 'expr' string value
+In boolean expressions we refer to the checks present in same profile and evaluate the result based on the result of those checks.
+A boolean expressio is evaluated in the end after all other checks have been evaluated.
+While evaluating boolean expression the result of checks referred is taken into the account and no check is actually run.
+We can provide a string in expression which consists of following expressions:
+AND, OR and NOT along with parenthesis to group the results
+"""
+import logging
+from hubblestack.utils.hubble_error import AuditCheckValidationError
+from pyparsing import infixNotation, opAssoc, Keyword, Word, alphas, ParserElement
+
+log = logging.getLogger(__name__)
+
+def execute(check_id, audit_check, result_list):
+    """
+    Execute single check
+    :param check_id: Unique check id
+    :param audit_check: Dictionary of an individual check implementation
+    :param result_list: List of results for other checks in profile
+    :return: dict -- dictionary of result status and output
+    """
+    log.debug('Executing bexpr module for check-id: %s' % (check_id))
+    keyword_list = ['AND', 'OR', 'NOT', '(', ')']
+    expression = audit_check['expr'].strip()
+    #Seperating keywords on the basis of space
+    expression = expression.replace('(', ' ( ')
+    expression = expression.replace(')', ' ) ')
+    #Parse the expression and evaluate the result
+    #Splitting the expression on the basis of spaces
+    expr_list = expression.split(" ")
+    expr_list = list(filter(None, expr_list)) #Filtering out empty spaces
+    referred_checks_list = []
+    referred_checks_result = {}
+    for expr in expr_list:
+        if expr.upper() not in keyword_list:
+            referred_checks_list.append(expr)
+
+    #Fetch the result of check from result list and store the result of referred checks
+    error = {}
+    for check in referred_checks_list:
+        check_found = False
+        for result in result_list:
+            if result.get('check_id', '') == check:
+                check_found = True
+                check_result =  result.get('check_result', '')
+                if check_result == "Success":
+                    referred_checks_result[check] = "True"
+                elif check_result == "Failure":
+                    referred_checks_result[check] = "False"
+                else:
+                    error[check] = "Unable to fetch result of referred check due to errors. Check: %s" % (check)
+                break
+        if not check_found:
+            error[check] = "Referred check: %s is not available. Please verify correct check is referred." % (check)
+    if error:
+        raise AuditCheckValidationError(error)
+
+    #Convert the expression now in the format to be parsed by pyparsing module
```

Comment:
This code actually is confusing in between.
We can write a method in this file, and do this processing there (the one required by 3rd party parsing module.)

Reply:
Created a separate function for expression evaluation